ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/TypeContext.java

```diff
@@ -0,0 +1,58 @@
+package io.confluent.ksql.execution.codegen;
+
+import io.confluent.ksql.schema.ksql.types.SqlType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TypeContext {
+  private SqlType sqlType;
+  private final List<SqlType> inputTypes = new ArrayList<>();
+  private final Map<String, SqlType> lambdaTypeMapping = new HashMap<>();
+
+  public SqlType getSqlType() {
+    return sqlType;
+  }
+
+  public void setSqlType(final SqlType sqlType) {
+    this.sqlType = sqlType;
+  }
+
+  public List<SqlType> getInputTypes() {
+    if (inputTypes.size() == 0) {
+      return null;
+    }
+    return inputTypes;
+  }
+
+  public void addInputType(final SqlType inputType) {
+    this.inputTypes.add(inputType);
+  }
+
+  public void mapInputTypes(final List<String> argumentList) {
```

Comment:
how will this be used? it's generally better to avoid side-effects (changing state) if we can just return the type mapping at this point. if that's unavoidable, then that's OK

Reply:
It's used inside of the `VisitLambdaExpression` in `ExpressionTypeManager`, `CodeGenRunner` and `SqlToJavaVisitor` to associate the given lambda input variables (x, y, k, v, etc) with their SqlType. The mapping is referenced in `VisitLambdaLiteral`, so I think we do want it stored in the context for future use, rather than just returning the input type at this point

Reply:
On a side note: I gave some thoughts on that while reading the other PR, trying to see if we can just maintain one collection instead two (a list, and a map). But cannot come up with a solution that is as efficient as now. The main "nit" concern is that if should try to enforce "when the list of input types should not be appended any more" and "the mapping (this function call) should only be triggered once the list becomes immutable" -- today they are not enforced and hence our code may be vulnerable to bugs. If you have a better idea to keep it more restricted, that'd be appreciated :)

Reply:
Hmm... Right now the `notallInputsSeen` is the gatekeeper for adding inputs. The issue is that we only know that we're done seeing inputs once we get `false` from that function, which would happen after we hit `visitLambdaExpression` and do the mapping... If all lambda functions accepted the same number of inputs it would be fairly straightforward to make the list immutable, but without knowing how many inputs to expect I can't think of a good way to make it more restricted. We can make the map immutable after we add everything, but with multiple lambdas of different names but the same input type (like we talked about on the other pr), the map may need to be wiped and remapped when we see a new lambda. I don't have any other ideas right now but I can keep thinking

Reply:
Sounds good; fair enough :)